### PR TITLE
[jobframework] Deployment integration

### LINKED ISSUE #2717: MVP support / extension of support for serving workloads
**What would you like to be added**:

I would like to make sure we have basic support for running serving workloads for the use case of running AI inference.
In particular I would like to have support for Deployments, StatefulSets, and LeaderWorkerSets. 

In the MVP work the integrations are based on single Plain pods (for Deployments) or Pod Groups (for StatefulSets).

1. *Deployments* 
This is a follow up to https://github.com/kubernetes-sigs/kueue/issues/2677.

What is needed:
- introduce a dedicated Deployment integration, and validate that it can only be enabled when pod integration is enabled
- copy the queue-name from Deployment down to PodTemlates

2. *StatefulSets*

What is needed:
- introduce a dedicated StatefulSet integration, and validate that it can only be enabled when pod integration is enabled
- copy the queue-name from StatefulSet down to PodTemlates
- set the PodTemaplate labels for the PodGroup:
  
* kueue.x-k8s.io/queue-name - from STS
* kueue.x-k8s.io/pod-group-name - STS_ + STS name (+ probably some hash to avoid collisions as for workloads)
* kueue.x-k8s.io/pod-group-total-count - STS replica count 

In the longer run to support scaling of stateful sets we may need to do https://github.com/kubernetes-sigs/kueue/issues/77, but this is out of scope for the issue,

3. LeaderWorkerSet support is moved to a dedicated issue: https://github.com/kubernetes-sigs/kueue/issues/3232

**Why is this needed**:

To support use cases of running AI training and inference in the same clusters, where the access to GPU is constrained by Kueue.

**Completion requirements**:

The API changes required are minimal (just potentially new labels / annotations), so I believe a new KEP is not required, but we need a proper documentation.

This enhancement requires the following artifacts:

- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by vladikkuzn**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds deployment integration to job framework

#### Which issue(s) this PR fixes:

Fixes #2717

#### Does this PR introduce a user-facing change?

```release-note
Add integration for Deployment, where each Pod is treated as a separate Workload.
```

### CODE REVIEW

File: pkg/controller/jobs/noop/noop_controller.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func (r Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	concurrency := mgr.GetControllerOptions().GroupKindConcurrency[gvk.GroupKind().String()]
+	ctrl.Log.V(3).Info("Setting up Noop reconciler", "concurrency", concurrency)
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: concurrency,
+		}).
+		Complete(r)
```
Comment (trasc) [inline pkg/controller/jobs/noop/noop_controller.go:35]:
```suggestion
	return nil
```
I think it's safe.

File: pkg/controller/jobs/noop/noop_controller.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+var (
+	_   jobframework.JobReconcilerInterface = (*Reconciler)(nil)
+	gvk                                     = corev1.SchemeGroupVersion.WithKind("Noop")
```
Comment (trasc) [inline pkg/controller/jobs/noop/noop_controller.go:18]:
```suggestion
```

File: pkg/controller/jobs/noop/noop_controller.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+type Reconciler struct {
+}
+
+func (r Reconciler) Reconcile(_ context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func (r Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	concurrency := mgr.GetControllerOptions().GroupKindConcurrency[gvk.GroupKind().String()]
+	ctrl.Log.V(3).Info("Setting up Noop reconciler", "concurrency", concurrency)
```
Comment (trasc) [inline pkg/controller/jobs/noop/noop_controller.go:30]: The concurrency of doing nothing is not very relevant. We can make the `Reconciler` hold some kind of info about the type for which the noop reconciler is "started" and log something like 

`"Setting up  Noop reconciler for %q ", r.gvk `

File: pkg/controller/jobs/deployment/deployment_controller.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+func (d *Deployment) Object() client.Object {
+	return (*appsv1.Deployment)(d)
+}
+
+func (d *Deployment) GVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func SetupIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return jobframework.SetupWorkloadOwnerIndex(ctx, indexer, gvk)
```
Comment (trasc) [inline pkg/controller/jobs/deployment/deployment_controller.go:67]: I'm not sure we need this

File: pkg/controller/jobs/deployment/deployment_controller.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/controller/jobs/noop"
+)
+
+var (
+	gvk = corev1.SchemeGroupVersion.WithKind("Deployment")
```
Comment (trasc) [inline pkg/controller/jobs/deployment/deployment_controller.go:34]:
```suggestion
	gvk = appsv1.SchemeGroupVersion.WithKind("Deployment")
```

File: pkg/controller/jobs/noop/noop_controller.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: concurrency,
+		}).
+		Complete(r)
+}
+
+func NewReconciler(client client.Client, record record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	return &Reconciler{}
+}
```
Comment (trasc) [inline pkg/controller/jobs/noop/noop_controller.go:40]:
```suggestion
func NewReconcilerFactory(gvk ...) func (client client.Client, record record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
	return func (client client.Client, record record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
		return &Reconciler{gvk: gvk}
	}
}
```

File: pkg/controller/jobs/deployment/deployment_controller.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+const (
+	FrameworkName = "deployment"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:   SetupIndexes,
+		NewReconciler:  noop.NewReconciler,
```
Comment (trasc) [inline pkg/controller/jobs/deployment/deployment_controller.go:44]:
```suggestion
		NewReconciler: NewReconcilerFactory(gvk) ,
```

File: pkg/controller/jobs/deployment/deployment_webhook.go
Comment (trasc) [inline pkg/controller/jobs/deployment/deployment_webhook.go:1]: I think we should make `...Labels[constants.QueueLabel] ` imitable for both the Deployment and  Deployment.Spec.Template

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Add integration for Deployment, where each Pod is treated as a separate Workload.
```

File: apis/config/v1beta1/configuration_types.go [hunk 8]
```diff
@@ -344,6 +345,13 @@ type PodIntegrationOptions struct {
 	PodSelector *metav1.LabelSelector `json:"podSelector,omitempty"`
 }
 
+type DeploymentIntegrationOptions struct {
```
Comment (alculquicondor) [inline apis/config/v1beta1/configuration_types.go:348]: I don't see this struct being used.

File: cmd/kueue/main.go [hunk 9]
```diff
@@ -281,6 +282,7 @@ func setupControllers(mgr ctrl.Manager, cCache *cache.Cache, queues *queue.Manag
 		jobframework.WithLabelKeysToCopy(cfg.Integrations.LabelKeysToCopy),
 		jobframework.WithCache(cCache),
 		jobframework.WithQueues(queues),
+		jobframework.WithDefaultWebhookFactory(noop.NewSetupWebhookFactory),
```
Comment (alculquicondor) [inline cmd/kueue/main.go:285]: why does this need to be an option?

File: pkg/controller/jobframework/setup.go [hunk 10]
```diff
@@ -96,8 +94,13 @@ func (m *integrationManager) setupControllers(mgr ctrl.Manager, log logr.Logger,
 				return nil
 			}
 		}
-		if err := noop.SetupWebhook(mgr, cb.JobType); err != nil {
-			return fmt.Errorf("%s: unable to create noop webhook: %w", fwkNamePrefix, err)
+		if options.DefaultWebhookFactory == nil {
```
Comment (alculquicondor) [inline pkg/controller/jobframework/setup.go:97]: when is this used?

File: cmd/kueue/main.go [hunk 11]
```diff
@@ -281,6 +282,7 @@ func setupControllers(mgr ctrl.Manager, cCache *cache.Cache, queues *queue.Manag
 		jobframework.WithLabelKeysToCopy(cfg.Integrations.LabelKeysToCopy),
```
Comment (alculquicondor) [inline cmd/kueue/main.go:282]: It seems like you are missing something like: 

```golang
jobframework.WithIntegrationOptions(appsv1.SchemeGroupVersion.WithKind("Deployment").String(), cfg.Integrations.DeploymentOptions)
```

File: pkg/controller/jobs/deployment/deployment_webhook.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create deployment selector: %w", err)
+	}
+	if !deploymentSelector.Matches(labels.Set(d.Labels)) {
+		return nil
+	}
+
+	// Get deployment namespace and check for namespace label selector match
+	ns := &corev1.Namespace{}
+	if err := wh.client.Get(ctx, client.ObjectKey{Name: d.Namespace}, ns); err != nil {
```
Comment (alculquicondor) [inline pkg/controller/jobs/deployment/deployment_webhook.go:90]: we can skip this get if there is no namespace selector configured.

File: pkg/controller/jobs/noop/noop_controller.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+type Reconciler struct {
+	gvk schema.GroupVersionKind
+}
+
+func (r Reconciler) Reconcile(_ context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func (r Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ctrl.Log.V(3).Info("Setting up Noop reconciler", "gvk", r.gvk)
```
Comment (alculquicondor) [inline pkg/controller/jobs/noop/noop_controller.go:28]:
```suggestion
	ctrl.Log.V(3).Info("Skipped reconciler setup", "gvk", r.gvk)
```

File: pkg/controller/jobs/noop/noop_webhook.go [hunk 14]
```diff
@@ -22,19 +22,22 @@ import (
...
 
-// SetupWebhook configures the webhook for batchJob.
-func SetupWebhook(mgr ctrl.Manager, apiType runtime.Object) error {
-	wh := &webhook{}
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(apiType).
-		WithDefaulter(wh).
-		WithValidator(wh).
-		Complete()
+func NewSetupWebhookFactory(apiType runtime.Object) func(mgr ctrl.Manager, opts ...jobframework.Option) error {
```
Comment (alculquicondor) [inline pkg/controller/jobs/noop/noop_webhook.go:32]: It looks like an unnecessary change

File: pkg/controller/jobs/noop/noop_controller.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+var (
+	_ jobframework.JobReconcilerInterface = (*Reconciler)(nil)
```
Comment (alculquicondor) [inline pkg/controller/jobs/noop/noop_controller.go:16]: Remove this line to avoid the circular dependency

File: pkg/controller/jobs/deployment/deployment_controller.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+const (
+	FrameworkName = "deployment"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:   SetupIndexes,
+		NewReconciler:  noop.NewReconcilerFactory(gvk),
```
Comment (alculquicondor) [inline pkg/controller/jobs/deployment/deployment_controller.go:43]: I see, so we are blocked on the circular dependency because of this line. Let me think about it.

**Review by alculquicondor — COMMENTED** [review]
There is only one alternative I could find: move the noop webhook and reconciler into the framework package.

I think that makes sense... it's part of the building blocks for writing controllers.

**Comment by alculquicondor** [comment]
> Right now there's no circular dependency, and noop webhook can be used for any controller/object, so I think we should be good here

Yes, but we have this option that shouldn't be an option `jobframework.WithDefaultWebhookFactory`. It should just be fixed.

File: cmd/kueue/main.go [hunk 17]
```diff
@@ -279,6 +280,7 @@ func setupControllers(mgr ctrl.Manager, cCache *cache.Cache, queues *queue.Manag
 		jobframework.WithEnabledExternalFrameworks(cfg.Integrations.ExternalFrameworks),
 		jobframework.WithManagerName(constants.KueueName),
 		jobframework.WithLabelKeysToCopy(cfg.Integrations.LabelKeysToCopy),
+		jobframework.WithIntegrationOptions(appsv1.SchemeGroupVersion.WithKind("Deployment").String(), cfg.Integrations.DeploymentOptions),
```
Comment (alculquicondor) [inline cmd/kueue/main.go:283]: nit: keep it next to the other call to `.WithIntegrationOptions` a few lines above.

File: pkg/controller/jobframework/setup.go [hunk 18]
```diff
@@ -96,8 +94,9 @@ func (m *integrationManager) setupControllers(mgr ctrl.Manager, log logr.Logger,
 				return nil
 			}
 		}
-		if err := noop.SetupWebhook(mgr, cb.JobType); err != nil {
-			return fmt.Errorf("%s: unable to create noop webhook: %w", fwkNamePrefix, err)
+		webhookFactory := NewSetupWebhookFactory(cb.JobType)
+		if err := webhookFactory(mgr, opts...); err != nil {
+			return fmt.Errorf("%s: unable to create default webhook: %w", fwkNamePrefix, err)
```
Comment (alculquicondor) [inline pkg/controller/jobframework/setup.go:99]:
```suggestion
		if err := SetupNoopWebhook(mgr, cb.JobType); err != nil {
			return fmt.Errorf("%s: unable to create noop webhook: %w", fwkNamePrefix, err)
```

The indirection is no longer needed.

File: pkg/controller/jobframework/noop_controller.go [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+func (r Reconciler) Reconcile(_ context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func (r Reconciler) SetupWithManager(_ ctrl.Manager) error {
+	ctrl.Log.V(3).Info("Skipped reconciler setup", "gvk", r.gvk)
+	return nil
+}
+
+func NewReconcilerFactory(gvk schema.GroupVersionKind) ReconcilerFactory {
```
Comment (alculquicondor) [inline pkg/controller/jobframework/noop_controller.go:30]:
```suggestion
func NewNoopReconcilerFactory(gvk schema.GroupVersionKind) ReconcilerFactory {
```

File: pkg/controller/jobframework/noop_controller.go [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	_ JobReconcilerInterface = (*Reconciler)(nil)
+)
+
+type Reconciler struct {
```
Comment (alculquicondor) [inline pkg/controller/jobframework/noop_controller.go:17]:
```suggestion
type NoopReconciler struct {
```

File: pkg/util/testingjobs/deployment/wrappers.go [hunk 21]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2023 The Kubernetes Authors.
```
Comment (alculquicondor) [inline pkg/util/testingjobs/deployment/wrappers.go:2]:
```suggestion
Copyright 2024 The Kubernetes Authors.
```

**Comment by kannon92** [comment]
Should we add e2e tests on this PR or follow up?

**Comment by alculquicondor** [comment]
The integration is so simple, that E2E tests sound like overkill

**Comment by kannon92** [comment]
> The integration is so simple, that E2E tests sound like overkill

I see.

File: apis/config/v1beta1/configuration_types.go [hunk 22]
```diff
@@ -344,6 +347,13 @@ type PodIntegrationOptions struct {
...
```
Comment (tenzen-y) [inline apis/config/v1beta1/configuration_types.go:350]: @alculquicondor Is there any context the reason why we decided to introduce these selectors?

IIUS, we introduced the selectors for PodIntegration because the Pod will be created by various type controllers in spite of Kube official one like kube-controller-manager.

But, it seems that the Deployment use cases do not match the same reason with the Pod ones.
If we allow the Deployment integrations to add such selectors, adding selectors to all integrations like Job will be justified, right?

Comment (alculquicondor) [reply apis/config/v1beta1/configuration_types.go:350]: You are right, we didn't really discuss this.

In fact, we took the decision not to add this for Jobs, in favor of using ValidatingAdmissionPolicy https://github.com/kubernetes-sigs/kueue/issues/2119

@mimowo @mwielgus do you have any additional justification for adding this?

/hold

Comment (alculquicondor) [reply apis/config/v1beta1/configuration_types.go:350]: Or perhaps @vladikkuzn has some context that I missed

File: pkg/controller/jobs/deployment/deployment_webhook.go [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+
+func (wh *Webhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldDeployment := FromObject(oldObj)
+	newDeployment := FromObject(newObj)
+
+	log := ctrl.LoggerFrom(ctx).WithName("deployment-webhook").WithValues("deployment", klog.KObj(newDeployment))
+	log.V(5).Info("Validating update")
+	allErrs := apivalidation.ValidateImmutableField(
+		newDeployment.GetLabels()[constants.QueueLabel],
+		oldDeployment.GetLabels()[constants.QueueLabel],
```
Comment (tenzen-y) [inline pkg/controller/jobs/deployment/deployment_webhook.go:96]: I guess that this has a potential null pointer error since the `GetLabels` function will return nil if the Deployment does not have any labels, right?

https://github.com/kubernetes/apimachinery/blob/a8a2284d318b9213fca20b0058e969941ce1c90b/pkg/apis/meta/v1/meta.go#L163

File: pkg/controller/jobs/deployment/deployment_webhook_test.go [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+			},
+		},
+	}
+
+	testCases := map[string]struct {
+		initObjects                []client.Object
+		deployment                 *appsv1.Deployment
+		manageJobsWithoutQueueName bool
+		namespaceSelector          *metav1.LabelSelector
+		deploymentSelector         *metav1.LabelSelector
```
Comment (tenzen-y) [inline pkg/controller/jobs/deployment/deployment_webhook_test.go:48]:
```suggestion
```

File: pkg/controller/jobs/deployment/deployment_webhook_test.go [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+func TestDefault(t *testing.T) {
+	defaultNamespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-ns",
+			Labels: map[string]string{
+				"kubernetes.io/metadata.name": "test-ns",
+			},
+		},
+	}
+
```
Comment (tenzen-y) [inline pkg/controller/jobs/deployment/deployment_webhook_test.go:42]: Is this still needed?

File: pkg/util/testingjobs/deployment/wrappers.go [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+// Obj returns the inner Deployment.
+func (d *DeploymentWrapper) Obj() *appsv1.Deployment {
+	return &d.Deployment
+}
+
+// Clone returns deep copy of the Deployment.
+func (d *DeploymentWrapper) Clone() *DeploymentWrapper {
+	return &DeploymentWrapper{Deployment: *d.DeepCopy()}
+}
+
```
Comment (tenzen-y) [inline pkg/util/testingjobs/deployment/wrappers.go:52]: I guess that this is never used anywhere.

File: pkg/controller/jobs/deployment/deployment_webhook.go
Comment (tenzen-y) [inline pkg/controller/jobs/deployment/deployment_webhook.go:1]: Additionally, could you simply implement webhook integration tests?
Indeed, this webhook controller-gen annotations have bugs, and we can not detect the bug in the current CI.

File: pkg/controller/jobs/deployment/deployment_webhook.go (hunk 23 above)
Comment (alculquicondor) [reply pkg/controller/jobs/deployment/deployment_webhook.go:96]: This is fine, though https://go.dev/play/p/pP1v9ALY6hh

Comment (tenzen-y) [reply pkg/controller/jobs/deployment/deployment_webhook.go:96]: Oh, there are no error even though that is null reference. It seems to be tricky.

Thank you for confirmation.

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+					NamespaceSelector: &metav1.LabelSelector{
+						MatchExpressions: []metav1.LabelSelectorRequirement{
+							{
+								Key:      "kubernetes.io/metadata.name",
+								Operator: metav1.LabelSelectorOpNotIn,
+								Values:   []string{"kube-system", "kueue-system"},
+							},
+						},
+					},
+				}),
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:72]:
```suggestion
```

We do not need this Pod selectors, right?

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+			var (
+				deployment *appsv1.Deployment
+				lookupKey  types.NamespacedName
+			)
+
+			ginkgo.BeforeEach(func() {
+				deployment = testingdeployment.MakeDeployment("deployment-with-queue-name", ns.Name).
+					Queue("user-queue").
+					Obj()
+				lookupKey = types.NamespacedName{Name: deployment.Name, Namespace: deployment.Namespace}
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:101]:
```suggestion
				lookupKey = client.ObjectKeyFromObject(deployment)
```

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+
+			ginkgo.It("Should inject queue name to pod template labels", func() {
+				gomega.Expect(k8sClient.Create(ctx, deployment)).Should(gomega.Succeed())
+
+				createdDeployment := &appsv1.Deployment{}
+				gomega.Eventually(func() error {
+					return k8sClient.Get(ctx, lookupKey, createdDeployment)
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+
+				fmt.Printf("createdDeployment: %v\n", createdDeployment.Spec.Template.Labels)
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:112]:
```suggestion
```

Is this debug code, right?

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+				lookupKey = types.NamespacedName{Name: deployment.Name, Namespace: deployment.Namespace}
+			})
+
+			ginkgo.It("Should inject queue name to pod template labels", func() {
+				gomega.Expect(k8sClient.Create(ctx, deployment)).Should(gomega.Succeed())
+
+				createdDeployment := &appsv1.Deployment{}
+				gomega.Eventually(func() error {
+					return k8sClient.Get(ctx, lookupKey, createdDeployment)
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:110]:
```suggestion
				gomega.Eventually(func(g gomega.Gomega) error {
					createdDeployment := &appsv1.Deployment{}
					g.Expect(k8sClient.Get(ctx, lookupKey, createdDeployment)).Should(gomega.Succeeded())
					g.Expect(createdDeployment.Spec.Template.Labels[constants.QueueLabel]).To(gomega.Equal("user-queue"),
					"Queue name should be injected to pod template labels")
				}, util.Timeout, util.Interval).Should(gomega.Succeed())
```

We can avoid exposing the `createdDeployment`.

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+
+			ginkgo.It("Should not allow to modify queue label", func() {
+				gomega.Expect(
+					k8sClient.Patch(
+						ctx,
+						deployment,
+						client.MergeFrom(testingdeployment.MakeDeployment("deployment-with-queue-name", ns.Name).
+							Queue("updated-queue").
+							Obj())),
+				).To(gomega.HaveOccurred())
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:127]: Instead of using the patch operation, could you get the Deployment and update the queue-label?

File: pkg/controller/jobs/deployment/deployment_webhook.go [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+
+func (wh *Webhook) Default(ctx context.Context, obj runtime.Object) error {
+	d := fromObject(obj)
+	log := ctrl.LoggerFrom(ctx).WithName("deployment-webhook").WithValues("deployment", klog.KObj(d))
+	log.V(5).Info("Applying defaults")
+
+	if d.Spec.Template.Labels == nil {
+		d.Spec.Template.Labels = make(map[string]string, 1)
+	}
+	d.Spec.Template.Labels[constants.QueueLabel] = d.Labels[constants.QueueLabel]
```
Comment (tenzen-y) [inline pkg/controller/jobs/deployment/deployment_webhook.go:66]: My last question is why do we need to enforce all Deployment to enqueue?

Is there any decisition?

**Review by tenzen-y — COMMENTED** [review]
Basically, lgtm

File: pkg/controller/jobs/deployment/deployment_webhook.go (hunk 32 above)
Comment (alculquicondor) [reply pkg/controller/jobs/deployment/deployment_webhook.go:66]: Right, if the Deployment doesn't define any queue-name, we shouldn't be adding the label to the pods.

Comment (tenzen-y) [reply pkg/controller/jobs/deployment/deployment_webhook.go:66]: That sounds good to me.
Thank you for the clarifications.

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 33]
```diff
@@ -0,0 +1,148 @@
...
+				deployment = testingdeployment.MakeDeployment("deployment-with-queue-name", ns.Name).
+					Queue("user-queue").
+					Obj()
+				lookupKey = client.ObjectKeyFromObject(deployment)
+			})
+
+			ginkgo.It("Should inject queue name to pod template labels", func() {
+				gomega.Expect(k8sClient.Create(ctx, deployment)).Should(gomega.Succeed())
+
+				gomega.Eventually(func(g gomega.Gomega) error {
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:97]:
```suggestion
				gomega.Eventually(func(g gomega.Gomega) {
```

We do not need to return errors.

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 34]
```diff
@@ -0,0 +1,148 @@
...
+					createdDeployment := &appsv1.Deployment{}
+					g.Expect(k8sClient.Get(ctx, lookupKey, createdDeployment)).Should(gomega.Succeed())
+					g.Expect(createdDeployment.Spec.Template.Labels[constants.QueueLabel]).
+						To(
+							gomega.Equal("user-queue"),
+							"Queue name should be injected to pod template labels",
+						)
+					return nil
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:105]:
```suggestion
```

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 35]
```diff
@@ -0,0 +1,148 @@
...
+			ginkgo.BeforeEach(func() {
+				deployment = testingdeployment.MakeDeployment("deployment-without-queue-name", ns.Name).
+					Obj()
+				lookupKey = client.ObjectKeyFromObject(deployment)
+			})
+
+			ginkgo.It("Should not inject queue name to pod template labels", func() {
+				gomega.Expect(k8sClient.Create(ctx, deployment)).Should(gomega.Succeed())
+
+				gomega.Eventually(func(g gomega.Gomega) error {
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:135]:
```suggestion
				gomega.Eventually(func(g gomega.Gomega) {
```

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+					createdDeployment := &appsv1.Deployment{}
+					g.Expect(k8sClient.Get(ctx, lookupKey, createdDeployment)).Should(gomega.Succeed())
+					g.Expect(createdDeployment.Spec.Template.Labels[constants.QueueLabel]).
+						To(
+							gomega.BeEmpty(),
+							"Queue name should not be injected to pod template labels",
+						)
+					return nil
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:143]:
```suggestion
```

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go [hunk 37]
```diff
@@ -0,0 +1,148 @@
...
+			err = serverVersionFetcher.FetchServerVersion()
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+			ctx, k8sClient = fwk.RunManager(cfg, managerSetup(
+				nil,
+				jobframework.WithManageJobsWithoutQueueName(false),
+				jobframework.WithKubeServerVersion(serverVersionFetcher),
+				jobframework.WithIntegrationOptions(
+					corev1.SchemeGroupVersion.WithKind("Pod").String(),
+					&configapi.PodIntegrationOptions{},
```
Comment (tenzen-y) [inline test/integration/controller/jobs/deployment/deployment_webhook_test.go:61]:
```suggestion
```

Is there any reason why we need to set empty PodIntegrationOptions?
If not, let's remove them.

**Review by tenzen-y — COMMENTED** [review]
Mostly lgtm.
Just some nits.

@alculquicondor Do you have any other comments?

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go (hunk 37 above)
Comment (alculquicondor) [reply test/integration/controller/jobs/deployment/deployment_webhook_test.go:61]: Can you just not call `WithIntegrationOptions`?

Comment (alculquicondor) [reply test/integration/controller/jobs/deployment/deployment_webhook_test.go:61]: I guess you can't because of this https://github.com/kubernetes-sigs/kueue/blob/243e572c441d0c1f305ced572151c77d6e1e9367/pkg/controller/jobs/pod/pod_webhook.go#L94

I guess we should change that function to work with an empty struct, but we can leave that for a follow up.

/lgtm
/approve

**Review by tenzen-y — COMMENTED** [review]
/lgtm
/approve

/hold cancel

File: test/integration/controller/jobs/deployment/deployment_webhook_test.go (hunk 37 above)
Comment (tenzen-y) [reply test/integration/controller/jobs/deployment/deployment_webhook_test.go:61]: PR was opened: https://github.com/kubernetes-sigs/kueue/pull/2945
